src/base/LruMap.h

```diff
@@ -148,50 +184,69 @@ LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
     if (memLimit() == 0)
         return false;
 
-    index.push_front(new Entry(key, t));
-    storage.insert(MapPair(key, index.begin()));
+    // do not store expired() entries because findEntry() ignores
+    if (ttl <= 0)
+        return false;
+
+    auto e = new Entry(key, t, ttl);
+    memUsed_ += e->memoryUsed();
+    index.push_front(e);
+
+    auto m = MapPair(key, index.begin());
+    memUsed_ += memoryUsedBy(m);
+    storage.insert(m);
 
     ++entries_;
-    return true;
+
+    // if over memory limit, start purging
+    trim();
```

Comment:
Please do not call `trim()` twice for one `add()`. The first (old) call was misplaced AFAICT. Adding another call just makes the already bad code worse. You already had to rewrite most of the old `add()` method. I recommend ignore what that method used to be and writing it correctly from scratch. You can remove any non-changes exposed by diff after that rewrite.

Reply:
done